Normalize input indexing map to identity to help fusion

okkwon opened the pull request:
Testing matmul + genericOp(add+transpose)

okkwon commented:
The failing case for Vulkan is supposed to be fixed by https://github.com/google/iree/pull/8836.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionUtils.cpp:
```diff
@@ -64,10 +59,14 @@ bool areLinalgOpsFusableUsingTileAndFuse(OpOperand &use) {
   AffineMap consumerIndexingMap = consumer.getTiedIndexingMap(&use);
-  if (!consumerIndexingMap.isIdentity()) return false;
-
+  if (allowTranspose) {
```
I dont think we need this. The interchange should be making this still identity.

MaheshRavishankar reviewed (changes requested):
Nice! Some minor comments.

okkwon commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,105 @@
+      return failure();
+
+    for (OpOperand *operand : genericOp.getInputOperands()) {
+      if (!areLinalgOpsFusableUsingTileAndFuse(*operand,
```
We may just need to normalize genericOps to have the identity indexing map if it does not egardless of the producer. Do you see any side effect?

MaheshRavishankar replied:
It can get into a loop, where you switch it w.r.t one producer and switch it again w.r.t to another. So need a check here that make sure the pattern applies only once. Checking for the first named op producer and making that identity should address this

okkwon replied:
I pushed a change to normalize the input indexing maps to identity. Thus, once the input indexing maps are identity, the pass does not revert so we won't have a loop.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,90 @@
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (genericOp.getNumInputs() == 0 || genericOp.getNumOutputs() == 0)
```
Dont think we need this check. You can replace this with just `genericOp.hasTensorSemantics()`

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,90 @@
+      return failure();
+
+    // check the generic op can be fused with the input operands.
+    auto indexingMap0 =
```
We need to tighten this up. We need to check 
1) If there is already a producer of this operation where the producer is a linalg named op and the indexing map is identity. If that is the case, do nothing (cause the assumption is that it will fuse with that producer)
2) If not, pick the first named op producer and make the indexing map identity w.r.t that.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,97 @@
+                                PatternRewriter &rewriter) const override {
+    if (!genericOp.hasTensorSemantics()) return failure();
+
+    if (genericOp.getNumInputs() != 1) return failure();
```
This is going to make this apply to a very small set of ops. We can remove this condition

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,109 @@
+
+      if (isa<linalg::GenericOp>(producer)) return failure();
+
+      if (producer->getNumResults() != 1) return failure();
```
I think for now we can drop this requirement.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,109 @@
+      auto producer = operand->get().getDefiningOp<linalg::LinalgOp>();
+      if (!producer) continue;
+
+      if (isa<linalg::GenericOp>(producer)) return failure();
```
I think we can drop this condition to. This is done post elementwise fusion. So any trivial fusion is already done. We can assume that now we are only looking for fusion using tile + fuse.

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,109 @@
+    // indexing maps for linalg named ops
+    SmallVector<AffineMap> indexingMaps;
+
+    for (auto operand : genericOp.getInputOperands()) {
```
The goal here is to group operations that can tile and fuse into one dispatch. So the indexing map is only relevant to look at when the producer is a LinalgOp. So I'd expect something like this

```
// Pattern needs to trigger only on  elementwise ops.
if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) return failure();

Optional<AffineMap> mapForInterchange = llvm::None;
for (auto opOperand : genericOp.getInputOperands()) {
  auto producer = opOperand->get().getDefiningOp<linalg::LinalgOp>();
  if (!producer) continue;
  
  auto indexingMap = genericOp.getTiedIndexingMap(opOperand);
  // This is already identity. Nothing to do.
  if (indexingMap.isIdentity()) return failure();
  // The map must be a permutation. If it is not then look for other operand.
  if (!indexingMap.isPermutation()) continue;

  if (!mapForInterchange) mapForInterchange = indexingMap;
}
if (!mapForInterchange) return failure();
auto interchange = llvm::to_vector(llvm::map_range(mapForInterchange->getResults(), [&](AffineExpr e) { return e.dyn_cast<AffineDimExpr>().getPosition(); }))

auto interchangedOp = intechangeGenericOp(rewriter, genericOp, intechange);
rewriter.replaceOp(genericOp, interchangedOp);
return success();
```

okkwon replied:
Your code transposes only based on the first indexing map. When there are multiple linalg ops as operand, what do we expect? My code gives up because without performance estimation of the multiple cases, it is unclear which indexing map to choose.

okkwon replied:
`interchangeGenericOp` actually does the op replacement.

MaheshRavishankar replied:
> Your code transposes only based on the first indexing map. When there are multiple linalg ops as operand, what do we expect? My code gives up because without performance estimation of the multiple cases, it is unclear which indexing map to choose.

The approach is, 
1) If there is already a linalg op that can be tiled and fused with the operation, then there is nothing to do.
2) If there is none, then picking any linalg operation is equivalent. The goal is to reduce the number of dispatches, so al options are "equally good" (at least for a first order of approximation). Its a heuristic, but is just the bare minimum heuristic.

MaheshRavishankar replied:
> `interchangeGenericOp` actually does the op replacement.

My bad. Missed that!

MaheshRavishankar reviewed (dismissed):
Not sure this works. I left a snippet of what I think should be the pattern (dont know if it works out of the box).
Also the pattern didnt seem to replace the original op. That means this change should not have any effect. Maybe worth checking that the patch actually reduces the number of dispatches.

MaheshRavishankar commented:
There are a few regressions. Something for CPU codegen folks to look at for later. For now we can land this under a flag.

okkwon commented:
Hmm. There must be a change in the upstream that affects the fusion. matmul + transpose is no longer fused even after the interchange. Looking into it.

okkwon commented:
The previous code didn't fuse dispatch regions, but it changes the input and output indexing maps to be identity and permutation, respectively. The change on the indexing maps results in transposed work group counts. The performance a transpose operation needs to be studied separately.

MaheshRavishankar commented:
> The previous code didn't fuse dispatch regions, but it changes the input and output indexing maps to be identity and permutation, respectively. The change on the indexing maps results in transposed work group counts. The performance a transpose operation needs to be studied separately.

Do you know why it didn't fuse dispatch regions. It should. If it doesn't then this transformation is unnecessary

okkwon commented:
> Do you know why it didn't fuse dispatch regions. It should. If it doesn't then this transformation is unnecessary

Yes, a code was missing in my PR. Added it back and now the fusion works as expected. Waiting for the benchmark results.

okkwon commented:
Significiant regressions were found from the arm benchmarks with the experimental config.

https://gist.githubusercontent.com/iree-github-actions-bot/eb5fbfc967ab9bc09056d4e793f7bd91/raw/8c6ef9d5058ab505af8adbb8f9ccb2a9831735c0/iree-full-benchmark-result-3017.md

Fusing mmt4d + transpose caused a change in the workgroup counts.

okkwon commented:
@hanhanW, would you mind reviewing this change? Mahesh is OOO. The feature is disabled by default, so it is not supposed to affect anything, but we can experiment it by toggling the flags.

hanhanW commented:
Yes, I can review the change. (I was looking for such support, good to see this. :D)

hanhanW commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,92 @@
+      if (!mapForInterchange) mapForInterchange = indexingMap;
+    }
+
+    if (!mapForInterchange) return failure();
```
Let's return `rewriter.notifyMatchFailure(op, "mismatch messages....")` with some messages?

hanhanW reviewed (commented):
Please add a test for `InterchangeTransposeGenericOpsPass`. I know there is a test that runs the pass + dispatch_linalg_on_tensors. It's better to also have a test which runs the pass only.

hanhanW commented:
note: I found that the interchanged transpose op is not vectorized when the flag is on. It would create a stack buffer because ConvertToDestinationPassingStyle + upstream bufferization does not work with the case. I sent out https://reviews.llvm.org/D126454 for fixing it.

okkwon commented:
Thanks @hanhanW for the review. I addressed your comments. It is great to see that you have a fix for the vectorization. I will do a perf run once the change appears in IREE.

hanhanW commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/test/normalize-input-indexing-map.mlir:
```diff
@@ -0,0 +1,52 @@
+// RUN: iree-opt --split-input-file --verify-diagnostics --iree-flow-interchange-transpose-generic-ops --canonicalize -cse %s | FileCheck %s
```
normalize-input-indexing-map.mlir -> interchange_transpose_generic_ops.mlir

I think the convention is that the test file name should match the pass name. And we should match the existing style -- use `_` instead.

hanhanW commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    // Pattern needs to trigger only on  elementwise ops.
+    if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) {
```
We don't have to do it in the PR. The comments states that it's matching element-wise operation. Are we going to support other cases?

There are the checks about elementwise trait upstream. Maybe we can refactor some of those code and reuse them here.

https://github.com/llvm/llvm-project/blob/0252357b3e1e8f6d3bc51ac6d7ac075842b2c956/mlir/lib/Dialect/Linalg/Transforms/Vectorization.cpp#L435-L448

okkwon replied:
Thanks for the info. I will do it in a later PR.

hanhanW commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp:
```diff
@@ -0,0 +1,97 @@
+    if (!mapForInterchange) {
+      return rewriter.notifyMatchFailure(genericOp, "no eligible operands");
+    }
+    // make the input indexing maps identity by interchanging.
```
```suggestion
    // Make the input indexing maps identity by interchanging.
```

hanhanW reviewed (approved):
Overall looks nice to me!

hanhanW commented:
Can you make the PR title and description more descriptive?

MaheshRavishankar commented on compiler/src/iree/compiler/Dialect/Flow/Transforms/FusionUtils.cpp:
```diff
@@ -27,9 +33,15 @@ namespace Flow {
   auto canTieWithOutsOperand = [&](OpOperand *outsOperand) {
-    if (linalgOp.getTiedIndexingMap(outsOperand) != insOperandIndexingMap) {
+    auto outputMap = linalgOp.getTiedIndexingMap(outsOperand);
+    if (clFuseTranspose) {
```
I am confused about this change. We should not need to do this. With the intechange pass, the dispatch region formation should fuse the ops directly (that is why we did the intechange). You need one or the other but not both. Can you give me an example where after interchange things still didnt fuse?

okkwon replied:
The transpose would have the following indexing maps:

input: (d0, d1) -> (d0, d1)
output: (d0, d1) -> (d1, d0)

Thus, the linalg op's input indexing map is different from the output indexing map, and the fusion is not triggered with the original code.

MaheshRavishankar replied:
Ah I see. Thanks! I'd actually hoist this option usage out of this method (this method is a hack which we might not need anymore, need to evaluate). So move this to line 79 and rename the flag

1) `iree-flow-fuse-transpose` -> `iree-flow-ensure-inplacable-consumer` . Make the new flag default true.
2) At line 79
```
return !clEnsureInplacable || canInsOperandTieWithOutsOperand
```

MaheshRavishankar replied:
Everything in this method (`canInsOperandTieWithOutsOperand`) is a WAR that wont work in the general case. So I am saying we add a flag to side-step this entire method for exploration of fusion with transpose-like ops (which is enabled by this change)
